google-cloud-spanner/src/main/java/com/google/cloud/spanner/MultiplexedSessionDatabaseClient.java

```diff
@@ -0,0 +1,428 @@
+    } catch (ExecutionException executionException) {
+      throw SpannerExceptionFactory.asSpannerException(executionException.getCause());
+    } catch (InterruptedException interruptedException) {
+      throw SpannerExceptionFactory.propagateInterrupt(interruptedException);
+    }
+  }
+
+  /**
+   * Returns true if the multiplexed session has been created. This client can be used before the
+   * session has been created, and will in that case use a delayed transaction that contains a
+   * future reference to the multiplexed session. The delayed transaction will block at the first
+   * actual statement that is being executed (e.g. the first query that is sent to Spanner).
+   */
+  private boolean isMultiplexedSessionCreated() {
+    return multiplexedSessionReference.get().isDone();
+  }
+
+  private DatabaseClient createMultiplexedSessionTransaction(boolean singleUse) {
+    Preconditions.checkState(!isClosed, "This client has been closed");
+    return isMultiplexedSessionCreated()
+        ? createDirectMultiplexedSessionTransaction(singleUse)
+        : createDelayedMultiplexSessionTransaction();
+  }
+
+  private MultiplexedSessionTransaction createDirectMultiplexedSessionTransaction(
+      boolean singleUse) {
+    try {
+      return new MultiplexedSessionTransaction(
+          this,
+          tracer.getCurrentSpan(),
+          // Getting the result of the SettableApiFuture that contains the multiplexed session will
+          // also automatically propagate any error that happened during the creation of the
+          // session, such as for example a DatabaseNotFound exception. We therefore do not need
+          // any special handling of such errors.
+          multiplexedSessionReference.get().get(),
+          singleUse ? getSingleUseChannelHint() : NO_CHANNEL_HINT,
+          singleUse);
+    } catch (ExecutionException executionException) {
+      throw SpannerExceptionFactory.asSpannerException(executionException.getCause());
+    } catch (InterruptedException interruptedException) {
+      throw SpannerExceptionFactory.propagateInterrupt(interruptedException);
+    }
+  }
+
+  private DelayedMultiplexedSessionTransaction createDelayedMultiplexSessionTransaction() {
+    return new DelayedMultiplexedSessionTransaction(
+        this, tracer.getCurrentSpan(), multiplexedSessionReference.get());
+  }
+
+  private int getSingleUseChannelHint() {
+    if (this.numCurrentSingleUseTransactions.incrementAndGet() > this.numChannels) {
+      return NO_CHANNEL_HINT;
+    }
+    synchronized (this.channelUsage) {
+      // Get the first unused channel.
+      int channel = this.channelUsage.nextClearBit(/* fromIndex = */ 0);
+      // BitSet returns an index larger than its original size if all the bits are set.
+      // This then means that all channels have already been assigned to single-use transactions,
+      // and that we should not use a specific channel, but rather pick a random one.
```

Comment:
`but rather pick a random one`

I am a bit confused by this comment. If we return NO_CHANNEL_HINT, we would pass an empty options map to GAX which will do round-robin? Do we need a correction to the docs here?

Reply:
See also my comment above.

This is why the code in `AbstractReadContext` is not dead. This line will generate the random channel hint: https://github.com/googleapis/java-spanner/blob/5979f2de26b9ec0a4c6a5f2c91b27304dcd9404a/google-cloud-spanner/src/main/java/com/google/cloud/spanner/AbstractReadContext.java#L195-L197

The reason for that is that by passing in `NO_CHANNEL_HINT` to the constructor of `SessionImpl`, then `SessionImpl` will use `sessionReference.getOptions()`. That returns `null` for multiplexed sessions, and the actually affiliated channel for pooled sessions. And if `options` is null, then the second argument that is passed in to `getChannelHintOptions` will be used (which is a random channel hint).

Reply:
Got it, thanks. I was confused by the `NO_CHANNEL_HINT`. What we mean is `NO_CHANNEL_HINT` returned from this layer. But ultimately it will always fallback to the random hint on `AbstractReadContext`.